### ISSUE TITLE
[question] Is there a way to override the default copy/paste functionality in DataGrid?

### ISSUE THREAD

**Issue by RyleySill93** [mui/mui-x#13895]
### Order ID or Support key 💳

75803

### Search keywords

copy paste data grid

### Latest version

- [X] I have tested the latest version

### The problem in depth

My understanding of the current copy/paste functionality is that only the value of each cell can be copy/pasted. What I'm looking to do is copy not only the cell value but also all the metadata associated with that cell (think formatting, cell ids, etc). 

My plan to do this is to use the `onClipboardCopy` prop to save the copied cell coordinates to a slice of internal state, and then use `onClipboardPasteEnd` to handle the logic of copying the state from my copied cell coordinates to the targeted cell coordinates. The problem is the default paste functionality tries to write values directly to the cell.

Slimmed down example here:
https://codesandbox.io/s/currying-leftpad-fh624t?file=/src/Demo.tsx

copy the "four" cell (which is bolded), paste into the "five" cell. Expectation is that the "five" cell would be replaced with a bold "four"

### Your environment

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by michelengelen** [comment]
Hey @RyleySill93 this is currently not possible.

I am not sure if we can build a supporting feature for this in the future. We would basically need to provide a possibility to override the copying function and pass it a selection of data so that the user can decide for themselves what is getting copied and what not. 

From what I read online the different clipboards should at least be able to hold a coplex data structure and if not it could at least be serialized into a stringified JSON or similar.

WDYT @mui/xgrid ?

**Comment by MBilalShafi** [comment]
> One last thing to note is wrapping my rows state setter with a window.setTimeout seems to fix the problem so it looks like this is a problem with the DataGrid not rerendering with the most up to date rows state?

@RyleySill93 It [works](https://codesandbox.io/p/sandbox/currying-leftpad-fh624t) with `setTimeout` because we [debounce](https://github.com/mui/mui-x/blob/5f8d458fd59e57c459e337cd37564392f54a5058/packages/x-data-grid-premium/src/hooks/features/clipboard/useGridClipboardImport.ts#L109) the row updates for them to be batched. So after you update the row in `onClipboardPasteEnd`, the `updateRows` function fires again and overwrites the existing value.
You could use a `setTimeout` with a value `> 50` as a _hacky_ workaround.

**Sidenote:** I also updated your code to use [the`updateRows` method](https://mui.com/x/react-data-grid/row-updates/#the-updaterows-method) instead of updating [the `rows` prop](https://mui.com/x/react-data-grid/row-updates/#the-rows-prop), as it's much better from the performance pov.

<hr />

Coming to the original problem, I agree with @michelengelen, we could certainly introduce a method similar to the `processRowUpdate` which would allow the users to transform the values before being copied, or perhaps enhance the `onClipboardCopy` to act as a transformer method.

**Comment by RyleySill93** [comment]
> @RyleySill93 It [works](https://codesandbox.io/p/sandbox/currying-leftpad-fh624t) with `setTimeout` because we [debounce](https://github.com/mui/mui-x/blob/5f8d458fd59e57c459e337cd37564392f54a5058/packages/x-data-grid-premium/src/hooks/features/clipboard/useGridClipboardImport.ts#L109) the row updates for them to be batched. So after you update the row in `onClipboardPasteEnd`, the `updateRows` function fires again and overwrites the existing value. You could use a `setTimeout` with a value `> 50` as a _hacky_ workaround, to run your row update logic after the Data Grid's internal one.

Ahh I see - this is definitely the source of my problem. I'm guessing the `updateRows` function is firing as a result of the paste action? This is the default behavior I'm looking to override - is there a way to disable it from firing on paste? 

Alternatively, is there a way to force the DataGrid to keep its internal rows state in sync with the rows prop?

> Coming to the original problem, I agree with @michelengelen, we could certainly introduce a method similar to the `processRowUpdate` which would allow the users to transform the values before being copied, or perhaps enhance the `onClipboardCopy` to act as a transformer method.

FWIW I'm not expecting MUI to handle the complex logic around copying meta data on the cell to the clipboard - I'm fine handling that logic in local state like I am in the linked sandbox. My issue is specifically around the DataGrid rows state failing to stay in sync with the `rows` prop.

**Comment by MBilalShafi** [comment]
> is there a way to disable it from firing on paste?

Yes, you could use [`onBeforeClipboardPasteEnd`](https://mui.com/x/react-data-grid/clipboard/#events) to cancel the default paste operation and handle it yourself instead.
Here's a codesandbox I tried to do it in: https://codesandbox.io/p/sandbox/zen-worker-3rt7vs

Could this work to solve your current use-case?

> Alternatively, is there a way to force the DataGrid to keep its internal rows state in sync with the rows prop?

> My issue is specifically around the DataGrid rows state failing to stay in sync with the rows prop.

This is a valid question indeed. The way `rows` prop works is not like any other controlled model. The reason I understand is its deep implications. Whenever there's a `rows` prop update, the Data Grid has to go through heavy computations like filtering, sorting, reordering, virtualization, row grouping, and so on, which are sometimes necessary to be done in a specific order for the Data Grid to work as expected, and it might happen in more than one render cycles

The Data Grid _does_ react on a `rows` prop update but it doesn't necessarily keep "in sync" with the rows prop like any other controlled model such as `filterModel` or `sortModel`. So when an internal update on the rows prop occurs, the value provided in `rows` prop gets "out of sync" with the Data Grid state.

@mui/xgrid This is indeed an interesting discussion, I'd like to know what other team members think on this topic.
What if we treat `rows` (and possibly `columns`) just like another controlled model, i.e. we keep it "in sync" with the rows prop, and on every internal state update or hydration, we call the method `onRowsUpdate`. The current behavior would be achievable with the new prop `initialState.rows`.

Do you see any drawbacks/anti-patterns to making this a default behavior in v8?

P.S: I'd be surprised if this hasn't been discussed/thought about before.
@flaviendelangle Any chance you can recall any thought given to making `rows` prop a controlled one while implementing https://github.com/mui/mui-x/issues/2044 and any reasons for not going that direction?
I am not expecting you to answer this for sure, since this is so much in the past and you aren't anymore working on the Data Grid, so feel free to ignore the mention 🙂

**Comment by flaviendelangle** [comment]
I can't recall us discussing this exact topic.
We had a lot of discussions about the lifecycle of the `rows` prop and the `updateRows` api method.
I would of course be cautious with those changes because they are at the heart of the grid logic and it's very easy to break important things, but no I don't have a absolute truth from 2020 to give here :laughing:

**Comment by cherniavskii** [comment]
Hi @RyleySill93 
How about using `processRowUpdate` instead of `onClipboardPasteEnd` + `setRows`?

Here is a demo: https://codesandbox.io/p/sandbox/condescending-artem-8lygxy
Is this the behavior you're looking for?
